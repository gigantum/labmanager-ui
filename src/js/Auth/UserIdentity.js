//vendor
import {
  graphql,
} from 'react-relay'
//mutations
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//environment
import {fetchQuery} from 'JS/createRelayEnvironment';

const userIdentityQuery = graphql`
  query UserIdentityQuery{
    userIdentity{
      id
      username
      email
      givenName
      familyName
      isSessionValid
    }
  }
`;

const UserIdentity = {
  getUserIdentity: () => {

    return new Promise((resolve, reject) =>{

      let fetchData = function(){

        fetchQuery(userIdentityQuery(), {}).then((response, error) => {

          if(response){
            resolve(response)
          }else{

            reject(response)
          }
        }).catch((error) =>{

          reject(error)
        })
      }

      fetchData()
    })
  },
  removeUserIdentity: () => {
    RemoveUserIdentityMutation(()=>{
      console.log('removed user')
    })
  }

}

export default UserIdentity
