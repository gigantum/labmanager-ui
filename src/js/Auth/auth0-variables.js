export const AUTH_CONFIG = {
  domain: 'calum.auth0.com',
  clientId: '9wFN5CZA4Z9xa1r8KZIZbevpjHG7bAab',
  callbackUrl: window.location.origin + '/callback'
}
