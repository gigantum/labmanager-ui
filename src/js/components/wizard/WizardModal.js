//vendor
import React from 'react'
import classNames from 'classnames'
//components
import CreateLabbook from './CreateLabbook'
import SelectBase from './SelectBase'
import TrackingToggle from './TrackingToggle'
import Loader from 'Components/shared/Loader'
import ButtonLoader from 'Components/shared/ButtonLoader'
import Modal from 'Components/shared/Modal'
//mutations
import CreateLabbookMutation from 'Mutations/CreateLabbookMutation'
import BuildImageMutation from 'Mutations/BuildImageMutation'
//store
import store from 'JS/redux/store'




export default class WizardModal extends React.Component {
  constructor(props){
  	super(props);

  	this.state = {
      'name': '',
      'description': '',
      'repository': '',
      'componentId': '',
      'revision': '',
      'selectedComponentId': 'createLabbook',
      'nextComponentId': 'selectBase',
      'previousComponentId': null,
      'continueDisabled': true,
      'modalBlur': false,
      'menuVisibility': true,
      'isTrackingOn': true,
      'createLabbookButtonState': ''
    };

    this._createLabbookCallback = this._createLabbookCallback.bind(this)
    this._createLabbookMutation = this._createLabbookMutation.bind(this)
    this._selectBaseCallback = this._selectBaseCallback.bind(this)
    this._continueSave = this._continueSave.bind(this)
    this._setComponent = this._setComponent.bind(this)
    this._hideModal = this._hideModal.bind(this)
    this._updateTextState = this._updateTextState.bind(this)
    this._setLabbookName = this._setLabbookName.bind(this)
    this._getSelectedComponentId = this._getSelectedComponentId.bind(this)
    this._toggleDisabledContinue = this._toggleDisabledContinue.bind(this)
    this._toggleMenuVisibility = this._toggleMenuVisibility.bind(this)
    this._setTracking = this._setTracking.bind(this)
  }
  /**
    @param {Object, string} evt,field
    updates text in a state object and passes object to setState method
  */
  _updateTextState = (evt, field) =>{
    let state = {}
    state[field] = evt.target.value;
    this.setState(state)
  }
  /**
    @param {bool} menuVisibility
    shows hides navigation menu
  */
  _toggleMenuVisibility(menuVisibility){
    this.setState({menuVisibility: menuVisibility})
  }

  _setTracking(trackingState){

    this.setState({
      isTrackingOn: trackingState
    })
  }

  /**
    @param {}
    shows modal window by update component state
  */
  _showModal = () => {
    this.setState({
      'modal_visible': true,
      'selectedComponentId': 'createLabbook',
      'nextComponentId': 'selectBase',
      'previousComponent': null
    })
  }
  /**
    @param {}
    hides modal window by update component state
  */
  _hideModal = () => {
    this.setState({'modal_visible': false, menuVisibility: true})
  }
  /**
    @param {string} componentId
    sets view for child components using and id
  */
  _setComponent = (componentId) => {

    this.setState({'selectedComponentId': componentId})

  }
  /**
    @param {string} labbookName
    sets labbookName for mini session
  */
  _setLabbookName = (labbookName) => {
    this.setState({'labbookName': labbookName})
  }

  /**
    @param {Object} base
    sets baseimage object for mini session
  */
  _setBase = (base) => {
    this.setState({'base': base})
  }

  /**
    @param {}
    gets id of current selected component for view navigation
    @return {string} selectedComponentId
  */
  _getSelectedComponentId = () => {
    return this.state.selectedComponentId
  }

  /**
    @param {boolean} isDisabled
    setsContinueDisabled value to true or false
  */
  _toggleDisabledContinue = (isDisabled) => {
    this.setState({
      'continueDisabled': isDisabled
    })
  }

  /**
    @param { boolean} isSkip
    gets selected id and triggers continueSave function using refs
  */
  _continueSave = ({isSkip, text}) =>{
    this.refs[this._getSelectedComponentId()].continueSave(isSkip);
    this.setState({'continueDisabled': true});
    if (text === 'Create Labbook') this.setState({ modalBlur: true })
  }
  /**
    @param {string ,string} name,description
    sets name and description to state for create labbook mutation
  */
  _createLabbookCallback(name, description){
    this.setState({
      name,
      description
    })

    this._setComponent('selectBase')
  }
  /**
    @param {string, string ,Int} repository, componentId revision
    sets (repository, componentId and revision to state for create labbook mutation
  */
  _selectBaseCallback(node){

    const {repository, componentId, revision} = node
    this.setState({
      repository: repository,
      componentId: componentId,
      revision: revision
    })
    //this._creatLabbookMutation();
  }
  /**
      @param {}
      sets name and description to state for create labbook mutation
  */
  _createLabbookMutation(){
    let self = this;
    const {
      name,
      description,
      repository,
      componentId,
      revision
    } = this.state

    this.setState({
      createLabbookButtonState: 'loading'
    })

    CreateLabbookMutation(
      name,
      description,
      repository,
      componentId,
      revision,
      !self.state.isTrackingOn,
      (response, error) => {
        if(error){


          store.dispatch({
            type: 'ERROR_MESSAGE',
            payload: {
              message: `An error occured while trying to create Project '${name}'.`,
              messageBody: error
            }
          })

          this.setState({
            modalBlur: false ,
            createLabbookButtonState: 'error'
          })

          setTimeout(()=>{
            this.setState({
              createLabbookButtonState: ''
            })
          },2000)

        }else{
          
          const {owner, name} = response.createLabbook.labbook

          this.setState({
            createLabbookButtonState: 'finished'
          })

          setTimeout(()=>{
            self._buildImage(name, owner)


            self.props.history.push(`../projects/${owner}/${name}`)

            this.setState({
              createLabbookButtonState: ''
            })
          },2000)



        }
      }
    )
  }
  /**
      @param {name, owner}
      builds docker iamge of labbook
  */
  _buildImage(name, owner){
    BuildImageMutation(
      name,
      owner,
      false,
      (response, error)=>{
        if(error){
          console.error(error)
          store.dispatch(
            {
              type: 'ERROR_MESSAGE',
              payload: {
                message: `ERROR: Failed to build ${name}`,
                messsagesList: error
            }
          })
        }
      })
  }

  render(){
    let loaderCSS = classNames({
      'hidden': !this.state.modalBlur
    })
    let currentComponent = this._currentComponent()
    return(
      <div>
        {
          this.state.modal_visible &&
          <Modal
            size="large"
            handleClose={() => this._hideModal()}
            header={currentComponent.header}
            preHeader={currentComponent.preHeader}
            renderContent={() =>
              <div>
                {
                  currentComponent.component
                }
                <ModalNav
                  self={this}
                  state={this.state}
                  getSelectedComponentId={this._getSelectedComponentId}
                  setComponent={this._setComponent}
                  hideModal={this._hideModal}
                  continueSave={this._continueSave}
                  createLabbookCallback={this._createLabbookCallback}
                />
              </div>
            }
          />
        }
        {
          this.state.modalBlur &&
          <Loader className={loaderCSS}/>
        }
      </div>
      )
  }

  _currentComponent(){
    switch(this._getSelectedComponentId()){
      case 'createLabbook':
        return {
          component:
            (<CreateLabbook
              ref="createLabbook"
              createLabbookCallback={this._createLabbookCallback}
              toggleDisabledContinue={this._toggleDisabledContinue}
              history={this.props.history}
              hideModal={this._hideModal}
              auth={this.props.auth}
            />),
          header: "Create Project"
        }

      case 'selectBase':
        return {
          component:
          (<SelectBase
            ref="selectBase"
            selectBaseCallback={this._selectBaseCallback}
            toggleDisabledContinue={this._toggleDisabledContinue}
            createLabbookMutation={this._createLabbookMutation}
            toggleMenuVisibility={this._toggleMenuVisibility}
          />),
          header: 'Select A Base',
          preHeader: 'Create Project'
        }
      default:
      return {
        component:
          (<CreateLabbook
            ref="createLabbook"
            createLabbookCallback={this._createLabbookCallback}
            toggleDisabledContinue={this._toggleDisabledContinue}
            history={this.props.history}
            hideModal={this._hideModal}
            auth={this.props.auth}
          />),
        header: "Create Project"
      }
    }
  }

}

/**
  @param {}
  gets button text for current componenet
  @return {string} text
*/
function ModalNav({self, state, getSelectedComponentId, setComponent, hideModal, continueSave}){

  let backButton = classNames({
    'WizardModal__progress-button': true,
    'button--flat': true,
    'hidden': (state.selectedComponentId === 'createLabbook')
  })

  let trackingButton = classNames({
    'hidden': (state.selectedComponentId !== 'createLabbook')
  })

  let wizardModalNav = classNames({
    'WizardModal__nav': true,
    'hidden': !state.menuVisibility
  })

  return(
    <div className={wizardModalNav}>
      <div>
        <button
          onClick={() => {setComponent('createLabbook')}}
          className={backButton}>
          Back
        </button>
        <div className={trackingButton}>
          <TrackingToggle
            setTracking={self._setTracking}
          />
        </div>
      </div>
      <div className="WizardModal__nav-group">
        <div className="WizardModal__nav-item">
          <button
            onClick={() => {hideModal()}}
            className="WizardModal__progress-button button--flat">
            Cancel
          </button>
        </div>

        <div className="WizardModal__nav-item">
          { (state.selectedComponentId === 'createLabbook') &&
            <button
              onClick={() => {continueSave({isSkip: false, text: 'Continue'})}}
              disabled={(state.continueDisabled)}>
              Continue
            </button>
          }

          { (state.selectedComponentId === 'selectBase') &&
            <ButtonLoader
              buttonState={state.createLabbookButtonState}
              buttonText={"Create Project"}
              className=""
              params={{isSkip: false, text: 'Create Project'}}
              buttonDisabled={state.continueDisabled}
              clicked={continueSave}
            />
          }
        </div>
      </div>
    </div>)
}
