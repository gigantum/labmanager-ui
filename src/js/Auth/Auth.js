import history from 'JS/history';
import auth0 from 'auth0-js';
import { AUTH_CONFIG } from './auth0-variables';
import RemoveUserIdentityMutation from 'Mutations/RemoveUserIdentityMutation'
//store
import store from 'JS/redux/store'


export default class Auth {
  auth0 = new auth0.WebAuth({
    domain: AUTH_CONFIG.domain,
    clientID: AUTH_CONFIG.clientId,
    redirectUri: AUTH_CONFIG.callbackUrl,
    audience:  AUTH_CONFIG.audience,
    responseType: 'token id_token',
    scope: 'openid profile email user_metadata'
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);

  }

  login() {
    this.auth0.authorize();
  }

  handleAuthentication() {

    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {

        this.setSession(authResult);

      } else if (err) {
        console.error(err);
      //  alert(`Error: ${err.error}. Check the console for further details.`);
      }
    });
  }

  setSession(authResult) {

    // Set the time that the access token will expire at
    let expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
    localStorage.setItem('access_token', authResult.accessToken);
    localStorage.setItem('id_token', authResult.idToken);
    localStorage.setItem('expires_at', expiresAt);
    localStorage.setItem('family_name', authResult.idTokenPayload.family_name);
    localStorage.setItem('given_name', authResult.idTokenPayload.given_name);
    localStorage.setItem('email', authResult.idTokenPayload.email);
    localStorage.setItem('username', authResult.idTokenPayload.nickname);
    //redirect to labbooks when user logs in

    let storrageRoute = sessionStorage.getItem('CALLBACK_ROUTE')
    let route = storrageRoute !== '' ? storrageRoute : `/labbooks`

    history.replace(route)
  }

  logout() {

    localStorage.removeItem('access_token');
    localStorage.removeItem('id_token');
    localStorage.removeItem('expires_at');
    localStorage.removeItem('family_name')
    localStorage.removeItem('given_name');
    localStorage.removeItem('email');
    localStorage.removeItem('username');
    RemoveUserIdentityMutation(()=>{
      //redirect to root when user logs out
      history.replace('/');
    })
  }

  isAuthenticated() {
    // Check whether the current time is past the
    // access token's expiry time
    let expiresAt = JSON.parse(localStorage.getItem('expires_at'));

    return new Date().getTime() < expiresAt;
  }
}
