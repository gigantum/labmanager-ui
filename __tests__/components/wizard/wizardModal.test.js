import React from 'react';
import WizardModal from 'Components/wizard/WizardModal';
import {mount} from 'enzyme'
import history from 'JS/history'
import renderer from 'react-test-renderer';
import Auth from 'JS/Auth/Auth';

let handler = () => {}

test('Test WizardModal rendering', () => {

  const component = renderer.create(
      <WizardModal
        history={history}
        handler={handler}
      />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
