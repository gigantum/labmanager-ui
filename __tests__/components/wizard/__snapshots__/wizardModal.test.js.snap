// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test WizardModal rendering 1`] = `<div />`;
